Cekees/pstab

alistairbntl opened the pull request.

cekees commented on proteus/LinearAlgebraTools.py:
```diff
@@ -231,10 +315,38 @@ def scatter_reverse_add(self):
 class ParVec_petsc4py(p4pyPETSc.Vec):
     """
     Parallel vector using petsc4py's wrappers for PETSc
+    ARB Question - I think this WIP has been completed.  Do you agree ckees? 
```
Agreed. This works for mixed spaces now too.

cekees replied:
The comment should be removed.

cekees commented on proteus/LinearSolvers.py:
```diff
@@ -719,18 +750,409 @@ def __init__(self, linear_smoother, pde_type='general_saddle_point'):
+    def initializeTwoPhaseCp_rho(self):
+        """Initialize a two phase scaled advection operator Cp.
+
+<<<<<<< HEAD
```
this appears to have unresolved conflicts in it.

cekees commented:
Looks like 7e10d80 was not a complete merge (it committed unresolved conflicts).

cekees commented on proteus/LinearAlgebraTools.py:
```diff
@@ -194,6 +202,82 @@ def petsc_create_diagonal_inv_matrix(sparse_petsc):
     diag_inv.setDiagonal(1./sparse_petsc.getDiagonal())
     return diag_inv
 
+# TODO - figure out whether this function is needed
```
This looks like a duplicate of the function of the same name below.

cekees commented on proteus/LinearAlgebraTools.py:
```diff
@@ -301,15 +413,85 @@ def scatter_reverse_add(self):
+    def print_info(cls):
+        import Comm
+        comm = Comm.get()
+        print 'comm.rank() = ' + `comm.rank()` + ' par_bs = ' + `cls.par_bs`
```
Use logEvent, not print. If you want to log on all processors then use `-A, --logAllProcesses`. There are several reasons 1) this version of print goes away in Python 3, 2) if somebody accidentally uses this on a large multi-processor job it can easily fill up their stdio buffer, 3) we want the log to contain all relevant information

cekees commented on proteus/LinearAlgebraTools.py:
```diff
@@ -301,15 +413,85 @@ def scatter_reverse_add(self):
+    proteus_jacobian : :class:`proteus.superluWrappers.SparseMatrix`
+        Jacobian generated by Transport class's initializeJacobian.
+    nzval_proteus2petsc : :class:`numpy.ndarray`
+        ?
```
?

cekees commented on proteus/LinearAlgebraTools.py:
```diff
@@ -345,6 +527,7 @@ def __init__(self,ghosted_csr_mat=None,par_bs=None,par_n=None,par_N=None,par_ngh
         comm = Comm.get()
         logEvent("ParMat_petsc4py comm.rank= %s blockSize = %s par_n= %s par_N=%s par_nghost=%s par_jacobian.getSizes()= %s "
                  % (comm.rank(),self.blockSize,par_n,par_N,par_nghost,self.getSizes()))
+#        import pdb ; pdb.set_trace()
```
remove

cekees commented on proteus/LinearAlgebraTools.py:
```diff
@@ -724,8 +983,166 @@ def apply(self,A,x,y):
+        self.kspSp.setOperators(self.Sp,self.Sp)
+        self.kspSp.setOptionsPrefix('innerSp_solve_')
+        self.kspSp.setFromOptions()
+        self.Sp.setNullSpace(self.const_null_space)
```
Can you explain the route to setting self.const_null_space? Which object has the responsibility of deciding if it has one or not?

cekees commented on proteus/LinearAlgebraTools.py:
```diff
@@ -724,8 +983,166 @@ def apply(self,A,x,y):
+    
+
+class TwoPhase_PCDInv_shell(InvOperatorShell):
+    r""" Shell class for the two-phase PCD preconditioner.  The
```
Thanks for putting the time into the doc string. It's good to use a consistent notation for these preconditioners.

cekees commented on proteus/LinearAlgebraTools.py:
```diff
@@ -724,8 +983,166 @@ def apply(self,A,x,y):
+
+        self._options = p4pyPETSc.Options()
+        self._create_constant_nullspace()
+        # TODO ARB - Need to implement the Chebyshev semi-iteration for
```
Didn't you already do that?

cekees commented on proteus/LinearAlgebraTools.py:
```diff
@@ -983,3 +1400,141 @@ def norm(self,y,type):
+# inputing the inverse diagonally preconditioned matrix A and b.  I'm
+# not sure this is the best approach, but I'd like to leave this code
+# in place for now in case it is useful in the future.
+
```
Makes sense.

cekees commented on proteus/LinearSolvers.py:
```diff
@@ -411,7 +417,11 @@ def __init__(self,L,par_L,
         self.csr_rep_local = self.petsc_L.csr_rep_local
         self.csr_rep = self.petsc_L.csr_rep
-        self.bdyNullSpace = bdyNullSpace
+        # HACK FOR CLOSED BDY PROBLEM - NEED TO CHANGE
```
We may need to put some more thought into this in general. The reason I put a link to the pde in the LinearOperator object is to pass information about the pde, specifically information on the operator's null space. This is going to be an issue for all the new work on projection schemes as well because we'll often be solving the pressure increment/correction poisson problem with a null space.

cekees commented on proteus/LinearSolvers.py:
```diff
@@ -523,37 +533,43 @@ def _defineNullSpaceVec(self,par_b):
-        pSpace = self.par_L.pde.u[0].femSpace
-        pressure_offsets = pSpace.dofMap.dof_offsets_subdomain_owned
-        n_DOF_pressure = pressure_offsets[rank+1] - pressure_offsets[rank]
-#        assert par_n == n_DOF_pressure
```
remove or uncomment?

cekees commented on proteus/LinearSolvers.py:
```diff
@@ -523,37 +533,43 @@ def _defineNullSpaceVec(self,par_b):
-        null_space_basis.scatter_forward_insert()
-        self.tmp_array[:] = self.tmp_array[petsc2proteus_subdomain]
-        self.global_null_space = [null_space_basis]
+#         par_N = par_b.getSize()
```
is this whole section a duplicate of the above?

cekees commented on proteus/LinearSolvers.py:
```diff
@@ -719,18 +750,433 @@ def __init__(self, linear_smoother, pde_type='general_saddle_point'):
             Currently supports Stokes and navierStokes
         """
+        import proteus
+        # TODO - find away to get rid of this import statement
```
Why is it necessary? Can't see it being used. Is this a testing issue?

cekees commented on proteus/MeshTools.py:
```diff
@@ -4837,6 +4837,37 @@ def refine(self):
         self.meshList.append(TriangularMesh())
         childrenDict = self.meshList[-1].refine(self.meshList[-2])
         self.elementChildren.append(childrenDict)
+    # def generateFromExistingCoarseMesh(self,mesh0,refinementLevels,nLayersOfOverlap=1,
```
this is implemented above. Remove this?

cekees commented on proteus/NumericalSolution.py:
```diff
@@ -270,13 +270,15 @@ def __init__(self,so,pList,nList,sList,opts,simFlagsList=None):
                     tmesh = TriangleTools.TriangleBaseMesh(baseFlags=n.triangleOptions,
                                                         nbase=1,
                                                         verbose=10)
+                    # ARB Before merge figure out the best way to deal with problem (eg in parallel
```
This is fixed by the simple change below, right? Just read the mesh on all processors so it is consisent?  We could raise an issue on refactoring the 2D mesh generation with triangle, gmsh, etc. and do this all better in the refactor.

cekees replied:
/cc @zhang-alvin

cekees commented on proteus/ShockCapturing.py:
```diff
@@ -44,6 +44,7 @@ def initializeElementQuadrature(self,mesh,t,cq):
                         cq[('da_sge',ci,ck,cj)]=cq[('da',ci,ck,cj)]
 
     def calculateNumericalDiffusion(self,q):
+        import pdb ; pdb.set_trace()
```
remove?

cekees commented on proteus/SplitOperator.py:
```diff
@@ -150,6 +150,57 @@ def setFromOptions(self,soOptions):
 
 Sequential_FixedStep = SO_base
 
+class Sequential_tnList(SO_base):
```
How does this differ from Sequential_FixedStep_Simple

alistairbntl replied:
As I understand the logic, of `Sequential_FixedStep_Simple`, `self.dt_system` is first set based on the length of the initial time step, which works fine.  
For subsequent time steps, `self.dt_system` starts at the value used in the previous iteration.  Thus, in `Sequential_FixedStep_Simple`'s  `stepExact_system` function (which uses the version written in the `SO_base` base class, see - https://github.com/erdc/proteus/blob/master/proteus/SplitOperator.py#L57 ), if `self.dt_system` is not large enough to reach the next time step, `self.dt_system` is set to step half the distance to `tExact`.  
In the case of wanting to replicate `tnList` exactly, it seems to me that we would want to set `self.dt_system` to give the exact step.  Thus, the difference is that `Sequential_tnList` sets `self.dt_system` to arrive at the next element of `tnList` in all cases (see - https://github.com/erdc/proteus/blob/c3159a489aae5f15266010b53feb4e355bc26017/proteus/SplitOperator.py#L167 )

cekees replied:
Got it. I don't think there's any reason to have both versions, so I removed the definition of Sequential_FixedStep_Simple and made it an alias of Sequential_tnList so it doesn't break anything.

cekees commented on proteus/SubgridError.py:
```diff
@@ -1506,8 +1506,6 @@ def __init__(self,coefficients,nd):
             coefficients.stencil[3].add(3)
     def calculateSubgridError(self,q):
         if self.nd == 2:
-        #    import pdb
```
remove?

cekees commented on proteus/Transport.py:
```diff
@@ -6577,11 +6569,12 @@ def initialize(self,
-                        assert (proteus_a[petsc2proteus_subdomain,:][:,petsc2proteus_subdomain] == petsc_a).all()
-                        assert((proteus_a == petsc_a[proteus2petsc_subdomain,:][:,proteus2petsc_subdomain]).all())
-                        petsc_a = np.arange(transport.dim**2).reshape(transport.dim,transport.dim)
+                                # TODO - ARB: these asserts need to be changed to reflect new petsc_a and proteus_a structures
```
fix or raise issue and do for next release?

cekees commented on proteus/mprans/RANS2P.py:
```diff
@@ -391,6 +400,56 @@ def attachModels(self,modelList):
         if self.epsFact_solid is None:
             self.epsFact_solid = numpy.ones(self.model.mesh.elementMaterialTypes.max()+1)
         assert len(self.epsFact_solid) > self.model.mesh.elementMaterialTypes.max(), "epsFact_solid  array is not large  enough for the materials  in this mesh; length must be greater  than largest  material type ID"
+        arbTest = True
```
This should be refactored to allow passing in a function for phi and vof. We do this in a lot of other places. @manuel-quezada needs this for RANS3P as well.

cekees commented on proteus/transportCoefficients.c:
```diff
@@ -2425,6 +2425,7 @@ void TwophaseNavierStokes_LS_SO_2D_Evaluate(const int nPoints,
   for (k=0;k<nPoints;k++)
     {
       /** \todo optimize coefficients */
+      printf("Here1\n");
```
remove

cekees commented on proteus/transportCoefficients.c:
```diff
@@ -2461,7 +2462,7 @@ void TwophaseNavierStokes_LS_SO_2D_Evaluate(const int nPoints,
       dmom_v_adv_v[k*2+0]=u[k];
       dmom_v_adv_v[k*2+1]=2.0*v[k];
-
+      printf("Here2\n");
```
remove

cekees commented on setup.py:
```diff
@@ -451,6 +451,10 @@
+                  (os.path.join(proteus_install_path,'tests','solver_tests','import_modules'),
+                   ['proteus/tests/solver_tests/import_modules/quad_mass_matrix.npy',
+                    'proteus/tests/solver_tests/import_modules/sol_10.npy',
+                    'proteus/tests/solver_tests/import_modules/sol_20_lst.npy']),
```
in the future we want to track these binary files with git lfs instead of git. @zhang-alvin has done this fairly extensively.

alistairbntl commented:
@cekees - I think I've addressed most of your comments above and yes, I see this PR replacing #575. 

Also, I have not tested the non-conservative for for 3D so I'm a little worried about that part, and it certainly needs a second pair of eyes.

cekees commented:
I think we should test this on 2D and 3D problems in air-water-vv quite a bit before we merge this. I did find some mistakes in  RANS2P.h in the new NONCONSERVATIVE option and spent a good bit of time making sure RANS2P2D.h and RANS2P.h were synchronized and cleaned up where possible.  @alistairbntl could you check it on all the cases for the 2D results in the paper? I'm running on the 3D marin problem on my laptop, and will try this on some HPC systems.

cekees commented on proteus/mprans/RANS2P.h:
```diff
@@ -347,3137 +454,5157 @@ namespace proteus
-  public:
-    const int nDOF_test_X_trial_element;
-    CompKernelType ck;
+    int nSpace,
```
Why did you do this?

cekees commented:
The tests are passing on my machine and the 3D RANS2P code is running with both NONCONSERVATIVE=0 & 1. Just waiting for the travis testing to get straightened out.  I think it has to do with their queue's.

cekees commented:
Just to make sure I'm following: Now the LevelModel __init__ can take a bdyNullSpace option to flag if the boundary conditions will generate a null space. That option can be set with the flag `boundaryCreatesNullSpace` in the physics  module (defined in default_p). @manuel-quezada you will probably want to use that in PresInc.py.

manuel-quezada commented:
What I did was to create a flag called fixNullSpace in the Coefficients of
PresInc.py. It is set to False by default to keep the code to work as it
was by default. This is the cheap trick of making the 0-th DOF to be set to
zero strongly. We can do this or let the user provide some desired mean
value pressure and then we do the adjustment accordingly.


On Tue, Sep 19, 2017 at 9:39 AM, Chris Kees <notifications@github.com>
wrote:

> Just to make sure I'm following: Now the LevelModel *init* can take a
> bdyNullSpace option to flag if the boundary conditions will generate a null
> space. That option can be set with the flag boundaryCreatesNullSpace in
> the physics module (defined in default_p). @manuel-quezada
> <https://github.com/manuel-quezada> you will probably want to use that in
> PresInc.py.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/erdc/proteus/pull/598#issuecomment-330561043>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFGi7d9w8oxMhahw5JVrY5k3X6N-MHyaks5sj9IMgaJpZM4PZFph>
> .
>

alistairbntl commented:
First, I've re-run everything using the non-conservative form using the latest commit and I haven't come across any issues yet.  I'm going to try  another 2d problem this afternoon, but so far so good.

Second, yes there is now a flag `bdyNullSpace` in the `LevelModel` class which is set in the _p file.  This bdyNullSpace attribute of the LevelModel is then set in KSP_Solver and used if needed.

alistairbntl commented:
Is PresInc solve a Laplacian type problem?

manuel-quezada commented:
Yes, it solves the Poisson problem with Neumann BCs so the null space has
the constant vector in it.

On Tue, Sep 19, 2017 at 10:42 AM, alistairbntl <notifications@github.com>
wrote:

> Is PresInc solve a Laplacian type problem?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/erdc/proteus/pull/598#issuecomment-330581111>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFGi7S_HeVvOqXKDQLk4U_yK-xddO-n8ks5sj-CzgaJpZM4PZFph>
> .
>

cekees commented:
@alistairbntl I've pushed my merged branch (baa56b7), so you're good to go.
